### PR TITLE
BUG: fix all RPATHs that start with $ORIGIN

### PR DESCRIPTION
**Pull request by metab0t**
Attempt to fix https://github.com/mesonbuild/meson-python/issues/509

### CODE REVIEW

**Review by dnicolodi — CHANGES_REQUESTED** [review]
Thanks. This looks good. Just one comment.

File: tests/packages/link-against-local-lib/ext/examplemod.c [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 #include <Python.h>
 
-#include "examplelib.h"
+#include "../src/examplelib.h"
```
[the remark this answers is not in the document]
Comment (metab0t) [reply tests/packages/link-against-local-lib/ext/examplemod.c]: Change the test accordingly.

File: mesonpy/_rpath.py [hunk 2]
```diff
@@ -27,8 +27,8 @@ def _set_rpath(filepath: Path, rpath: Iterable[str]) -> None:
 
     def fix_rpath(filepath: Path, libs_relative_path: str) -> None:
         rpath = _get_rpath(filepath)
-        if '$ORIGIN/' in rpath:
-            rpath = [('$ORIGIN/' + libs_relative_path if path == '$ORIGIN/' else path) for path in rpath]
+        if any(path.startswith('$ORIGIN/') for path in rpath):
+            rpath = ['$ORIGIN/' + libs_relative_path] + [path for path in rpath if not path.startswith('$ORIGIN/')]
```
Comment (dnicolodi) [inline mesonpy/_rpath.py:31]: The order of the entries in RPATH matters. This does not preserve it.

File: mesonpy/_rpath.py [hunk 3]
```diff
@@ -46,13 +46,21 @@ def _get_rpath(filepath: Path) -> List[str]:
...
-    def _replace_rpath(filepath: Path, old: str, new: str) -> None:
-        subprocess.run(['install_name_tool', '-rpath', old, new, os.fspath(filepath)], check=True)
+    def _replace_rpath(filepath: Path, old_rpaths: Iterable[str], new_rpaths: Iterable[str]) -> None:
+        command = ['install_name_tool']
+        for old in old_rpaths:
+            command.extend(['-delete_rpath', old])
+        for new in new_rpaths:
+            command.extend(['-add_rpath', new])
+        command.append(os.fspath(filepath))
+        subprocess.run(command, check=True)
```
Comment (dnicolodi) [inline mesonpy/_rpath.py:56]: Here too. Removing and adding back an RPATH does not preserve ordering.

**Comment by dnicolodi** [comment]
I've pushed a patch that fixes the identified issues. Please take a look.

**Comment by metab0t** [comment]
> I've pushed a patch that fixes the identified issues. Please take a look.

Thanks for the fix! LGTM

**Pull request merged**